### DISCUSSION TITLE
Mu should support connecting to BLE enabled [Circuit|Micro]Python devices

### DISCUSSION THREAD

**Discussion by ntoll** [mu-editor/mu#1333] [Ideas & New Features]
Mentioned by @tannewt given CircuitPython devices should soon have such capabilities and the micro:bit v2 should be able to support this (@carlosperate, am I right?).

A quick Google for Python packages for interfacing with BLE devices leads to [Bleak](https://bleak.readthedocs.io/en/latest/).

Please add ideas, UI suggestions and technical hiccups. ;-)

cc/ @kattni @DaveSprague @makermelissa  @fmorton

**Comment by dybber** [comment]
MicroPython also has a WebREPL interface that could be interesting to interface with, to program a MicroPython device over Wifi. I've thought about that for a while, but not had any good ideas for how to actually add it to the user-interface

**Reply by dybber** [reply]
Perhaps it could be added as a right-click menu on the device connectivity icon in the bottom-right corner? "Connect through BLE", "Connect through wifi"

**Reply by carlosperate** [reply]
I think right click features might be a bit hard to discover. Maybe a small rectangular button or two at the top of the REPL panel (the area between the text editor and the repl content) would be okay? but we would need to be careful about not adding too much stuff there.

**Reply by dybber** [reply]
Depends on whether we want it to be discoverable, or it's a niche feature that few will use. Currently you get an error if you try to open the REPL with no device attached, so that would need to change for your suggestion to work.

**Comment by tannewt** [comment]
Note, we already use Bleak underneath the Blinka version of bleio and I think it works reasonably well. @dhalbert would have a better idea.

**Comment by carlosperate** [comment]
@WayneKeenan (https://github.com/TheCellule/python-bleson) and @ukbaz (https://github.com/ukBaz/python-bluezero) also have a lot of experience with Python and Bluetooth on the desktop.  Do you have any recommendations? We would really appreciate your input.

**Reply by WayneKeenan** [reply]
I'd also have to say take a look at Bleak.  

I've only used it indirectly on macOS and Pi to talk to Lego Boost, but at least it's cross platform (the big 3) and actively maintained.

I did take a cursory glance under the hood at the time, I recall they have found a less verbose/simpler (than Bleson anyway) to bridge from Python to the Win BLE API, which is nice.   But I've not tried it.

**Reply by ukBaz** [reply]
I have read the heading of this a few times and I still remain uncertain what the desired behavior is that is being asked for. Is it:
 
1. to be able to program/flash the [circuit|micro] python device via Bluetooth from Mu on the desktop (e.g. similar to what you can do with the app: https://microbit.org/get-started/user-guide/ble-ios/) 
2. Use Mu to create the client and server sides of  connection. Again my assumption is the  [circuit|micro] python device would have the Peripheral GAP role and the desktop machine with Mu would be the Central role.

If I assume number 2 above, then is the question, what library should we recommend in Mu to create the client?

My efforts with Bluetooth has been focused on Linux so I am maybe not best placed to answer this but have a some insight. I keep some kind of list of Python libraries at: https://github.com/ukBaz/python-bluezero/wiki

For BLE central role there is only Bleak and pybluez that target being cross-platform like Mu: 

- pybluez on their github repo it says "This project is not under active development".
- Bleak makes use of `asyncio` which I am not sure if that is compatible with Mu's goal of "a simple Python editor for beginner programmers". I guess that is a broader conversation about what you can do with BLE synchronous vs asynchronously.

Just as Python supports both Linux and [Windows](https://bugs.python.org/issue36590) gets support for Bluetooth Classic Serial Port Profile (SPP) we all move our attention on to BLE  :-)

**Reply by WayneKeenan** [reply]
I think this isn't a user facing API request but a request to add a BLE transport alongside the Serial transport for (raw)REPL.   So, yeah, your #1

**Reply by carlosperate** [reply]
Thanks for your recommendation!

Yes, we are looking at having Mu stablish a BLE connection to the device (option 1).

Sounds like Bleak might be the best option.
Thanks for the asyncio warning, we'll definitely need to consider how to manage that. Taking in consideration we will likely need to spun out a new thread for this, we can hopefully write the BLE functionality almost as individual commands that are executed, left to do their own thing, and Mu just has to wait for them to finish, so we could probably let the event loop run and complete in their own process or thread.

**Reply by dhalbert** [reply]
As @tannewt mentioned, we use bleak as the middleware for our [adafruit-blinka-bleio](https://github.com/adafruit/Adafruit_Blinka_bleio) library that emulates the native BLE support we provide in CircuitPythno. Bleak is well-structured and well-maintained, and the authors are happy to consider outside PR's. I looked at several alternatives and it was clearly the best choice. It uses async/await, but I have a simple wrapper function that I wrote to create a synchronous-only interface. It acts only as a BLE central, not a peripheral. It works on MacOS, Linux, and Windows. Essentially all of our CircuitPython BLE libraries that expect to act as a central can work with it.

**Reply by carlosperate** [reply]
That's great, we'll definitely look at adafruit-blinka-bleio when it comes to implement this. Thanks Dan!
